Issue #11442: Check no closed issue references workflow is now triggered on pull requests

Vyom-Yadav opened the pull request:
Resolves #11442
Tested at - https://github.com/Vyom-Yadav/actions-test/pull/39
https://github.com/Vyom-Yadav/actions-test/pull/40

Direct push to master - https://github.com/Vyom-Yadav/actions-test/runs/5655841079?check_suite_focus=true

Push to PR (not from fork) - https://github.com/Vyom-Yadav/actions-test/runs/5655783382?check_suite_focus=true

Push to PR (from fork) - https://github.com/Vyom-Yadav/actions-test/runs/5655810498?check_suite_focus=true

---

What this PR does-

1. **Improved reporting:-** A clickable link is now given to the user which redirects it to the line where the closed issue is referenced. 
2. **Runs on a PR:-** The job now runs on a PR too.
3. **Issues linked to PR are mentioned:-** If the developer has linked some issue to the PR and that issue is referenced somewhere in the code then a violation will be given out to user. Rationale behind this is that if you [link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) an issue to a PR then when that PR is merged, then that issue would be automatically closed, linking issue is different than just mentioning it.

Vyom-Yadav commented:
The job will run on this PR but it will not pick up my shell script modifications as it uses the latest master-
https://github.com/checkstyle/checkstyle/blob/41eae30fe150ea4273f965fdbf4f7fcee150184a/.github/workflows/no_old_refs.yml#L21-L23

Vyom-Yadav commented:
Marked as a draft because if a PR is linked to an issue (may close that issue) and if it still references that issue then no error will be thrown, I am adding logic for it.

Vyom-Yadav commented:
Ready for review, a summary of what changes have been made-

- Better error reporting, with the file, line number, issue link.
- Workflow is now triggered on PR's
- Checks out to the current PR SHA
- If a PR is [linked to an issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) and that issue is in code somewhere (example- `until #112`) then an error will be thrown as that PR is supposed to close that issue and that issue is still referenced in code. If a PR doesn't fully close an issue then it shouldn't be linked to the issue.

rnveach commented:
> Better error reporting, with the file, line number, issue link.

Why don't we separate this and merge it faster. I don't see that part being an issue. This is also a good example of where we can split a PR into multiple. They both should reference the issue.

In this example, both are "linked" as you and github describe it, right?

rnveach commented:
> Workflow is now triggered on PR's

Should the workflow also run on an issue being closed to connect to times where we delay closing an issue after merge?

rnveach commented:
> ./config/pmd.xml:337 -> https://github.com/checkstyle/checkstyle/issues/11201

A thought, would it be better if the location was a github link? The CI automatically turns it into a clickable link for easy access.
Example: https://github.com/checkstyle/checkstyle/blob/master/config/pmd.xml#L337

Vyom-Yadav commented:
> Why don't we separate this and merge it faster. I don't see that part being an issue. This is also a good example of where we can split a PR into multiple. They both should reference the issue.

Yes, I never denied that they should reference that issue, but they should not be linked to it if it is done in multiple parts.

> It has to specifically use the keyword like fixes/closes/etc...?

Yes it has to use those particular words, or you have to manually link using the steps shown in https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

> Should the workflow also run on an issue being closed to connect to times where we delay closing an issue after merge?

I don't think it will be needed, it also runs when you merge the PR, so if there is something left, then merge commit will show it.

> Also, where is error reported if occurs in PR or it occurs on merge to master, or on close of issue if we extend it to these places?
> where is error reported if occurs in PR

The PR itself, there is a list of checks at the end.

> on merge to master

On merging it to master, the checks run again (if configured to do so)-
![Screenshot from 2022-03-20 21-15-53](https://user-images.githubusercontent.com/73882557/159170630-6d63b857-0b68-4e12-824e-bc212c5f4723.png)

> close of issue if we extend it to these places

They will run similarly to the image shown above, it will run on the latest commit to the default branch.

Vyom-Yadav commented:
> A thought, would it be better if the file location was a github link? The CI automatically turns it into a clickable link for easy access.

Yes, but the link will be to the users fork or master if committed directly.

rnveach commented:
Ok, yea, my confusion was on how github defines linked. I still considered the PR linked to the Issue even if it doesn't close it, but github doesn't see it that way.

> it also runs when you merge the PR, so if there is something left, then merge commit will show it

But this validation only occurs when the issue is closed, right? What happens if the PR is merged but the issue isn't closed until 24 hours later? There are a few PRs I pinged romani on to determine what is left to do for some outstanding issues that appeared done.

Vyom-Yadav commented:
> What happens if the PR is merged but the issue isn't closed until 24 hours later

When we merge a PR, we also make a push to the master branch, so the workflow will be triggered, there is also another way to do it using `pull_request_target` but the drawback is, it will be triggered when we close the PR without merging so first way is the way to go.

Vyom-Yadav commented:
```yaml
on:
  push:
    branches: [ master ]
  pull_request:
```
This type of workflow triggers on pushing anything to master branch or on a pull request (not on all PR events, by default only if it is opened or synchronized or reopened)

Vyom-Yadav commented:
> A thought, would it be better if the file location was a github link? The CI automatically turns it into a clickable link for easy access.

Did not edit the comment above, as a lot of comments are there in between.

We can make a link to the fork or to the master branch, both are possible, but if the user adds an until comment which is not valid, it won't be present in the master branch as those changes are only there in the fork.

strkkk commented:
I think action should only check for linked issue for PR run, not for all issues. In case master is failing, PR run will also fail, it can cause confusion for contributors.

Also, issue can be mentioned in several places, will you print them all? I do not see such cases in your test PR.
Initially I didn't print file names because you can just run search on issue number in codebase and quickly find all usages.

strkkk commented:
Please also add case when you have mentioned non-checkstyle issue with same id as mentioned checkstyle issue and vise versa, e.g.
mentioned <your_fork_link>/1234 and <checkstyle_link>/1234 not mentioned
not mentioned <your_fork_link>/1234 and <checkstyle_link>/1234 mentioned

romani commented:
Keep in mind we can reference issue in PR, but it is not resolve PR.

See as example of PR for message updates in inputs

romani commented on .ci/no_old_refs.sh:
```diff
@@ -5,27 +5,49 @@ set -e
+    printf "\nFollowing issues are mentioned in code to do something after they are closed:\n\n"
     cat $CLOSED_ISSUES
+    if [ -f "$LINKED_ISSUES_MENTIONED" ]; then
+      printf "\nFollowing issues are linked to the PR and are also mentioned in the code:\n\n"
```
Let's use echo
https://stackoverflow.com/a/10576796

Vyom-Yadav replied:
Done.

romani commented on .ci/no_old_refs.sh:
```diff
@@ -5,27 +5,49 @@ set -e
+fi
 
-for issue in $(sort -u $MENTIONED_ISSUES); do
+for line in $(sort -u $MENTIONED_ISSUES); do
```
MENTIONED_ISSUES_lines

rnveach replied:
Why not MENTIONED_ISSUES_LINES to be consistent with how we name things in Java?

Vyom-Yadav replied:
Done.

romani reviewed (changes requested):
Items:

Vyom-Yadav commented on .ci/no_old_refs.sh:
```diff
@@ -5,27 +5,49 @@ set -e
-grep -IPohr "(after|[Tt]il[l]?) $GITHUB_HOST/[\w.-]+/[\w.-]+/issues/\d{1,5}" . \
-  | sed -e 's/.*github.com\///' >> $MENTIONED_ISSUES
+grep -IPonr "(after|[Tt]il[l]?) $GITHUB_HOST/[\w.-]+/[\w.-]+/issues/\d{1,5}" . \
+  | perl -pe 's/:(?!\d).*github.com\//:/' >> $MENTIONED_ISSUES
```
No, `sed` doesn't support lookaheads, etc.

romani replied:
Why we need lookahead now ?

Vyom-Yadav replied:
There are multiple ways of doing it, don't know why I resorted to lookahead earlier, changed it to use a different regex supported by `sed`.

Vyom-Yadav commented:
> Keep in mind we can reference issue in PR, but it is not resolve PR.
See as example of PR for message updates in inputs

You can reference any issue in PR but If you link it to PR and merge that PR then that issue would be automatically closed, there is a difference in referencing and linking.
https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

romani commented:
Please put your testing updates to separate commit "minor: do not merge, it is just testing"

Vyom-Yadav commented:
> Please put your testing updates to separate commit "minor: do not merge, it is just testing"

Done, also don't worry about spelling failure, I will fix it, some words will be added to ignore list and some words like `Vyom` will be removed

romani commented on .ci/no_old_refs.sh:
```diff
@@ -5,27 +5,66 @@ set -e
 # collect checkstyle issues where only hash sign is used
-grep -IPohr "[Tt]il[l]? #\d{1,5}" . \
-  | sed -e 's/.*#/checkstyle\/checkstyle\/issues\//' >> $MENTIONED_ISSUES
+grep -IPonr "[Tt]il[l]? #\d{1,5}" . \
```
https://www.gnu.org/software/grep/manual/grep.html#index-_002dn
```
-n 
--line-number
Prefix each line of output with the 1-based line number within its input file. (-n is specified by POSIX.)
```

Why we need line numbers ?

romani replied:
https://www.gnu.org/software/grep/manual/grep.html#index-_002dh

```
-h
--no-filename
Suppress the prefixing of file names on output. This is the default when there is only one file (or only standard input) to search.
```

Vyom-Yadav replied:
https://github.com/checkstyle/checkstyle/runs/5686520746?check_suite_focus=true
Now CI reports a clickable link instead of manually telling where the closed issue is referenced.
https://github.com/Vyom-Yadav/checkstyle/blob/updateNoClosedRefWorkflow/src/main/java/com/puppycrawl/tools/checkstyle/checks/FinalParametersCheck.java#L241

In the link mentioned above, I need the filename (so `-h` was removed) and I need line number (#L241) so `-n` was added.

romani commented on .ci/no_old_refs.sh:
```diff
@@ -5,27 +5,66 @@ set -e
+  sed -i "s/^/$CHECKSTYLE_ISSUE_PREFIX/g" $LINKED_ISSUES_FORMATTED
+fi
+
+for MENTIONED_ISSUES_LINES in $(sort -u $MENTIONED_ISSUES); do
```
What for sorting is required ?

Vyom-Yadav replied:
It isn't required at all, previously it was required as we were just reporting the issue number and we do not wanted duplicates, no it isn't required, removed.

Vyom-Yadav replied:
It isn't required at all, previously it was required as we were just reporting the issue number and we do not wanted duplicates, no it isn't required, removed.

romani reviewed (changes requested):
Items:

romani commented:
Please update issue description to show exactly what you want to archive in this PR. You do bunch of good improvements but I am confused as i need to get from code what you want to archive.

It would help a lot if we confirm that written enhancements in issue matching to code in PR.

Vyom-Yadav commented:
@romani I updated the issue description to summarize the changes

romani commented:
I do not see any changes in issue description

romani commented on .ci/no_old_refs.sh:
```diff
@@ -5,27 +5,66 @@ set -e
+for MENTIONED_ISSUES_LINES in $(cat $MENTIONED_ISSUES); do
+  issue=${MENTIONED_ISSUES_LINES#*[0-9]:}
+
+  location=${MENTIONED_ISSUES_LINES%:[0-9]*}
```
Rename to file_path.


Please use can case for all variables. Looks like we follow that style.

Vyom-Yadav replied:
Done.

romani commented on .ci/no_old_refs.sh:
```diff
@@ -5,27 +5,66 @@ set -e
+    echo -ne "\nFollowing issues are mentioned in code to do something after they are closed:\n\n"
     cat $CLOSED_ISSUES
+    if [ -f "$LINKED_ISSUES_MENTIONED" ]; then
+      echo -ne "\nFollowing issues are linked to the PR and are also mentioned in the code:\n\n"
```
We need to hint user that link should be removed.

Vyom-Yadav replied:
Done.

romani reviewed (changes requested):
Items:

Vyom-Yadav commented:
> I do not see any changes in issue description

My bad, I had updated the PR description instead of the issue description, updated the issue description now.

romani commented on .ci/no_old_refs.sh:
```diff
@@ -5,27 +5,67 @@ set -e
+  sed -i "s/^/$CHECKSTYLE_ISSUE_PREFIX/" $LINKED_ISSUES_FORMATTED
+fi
+
+for MENTIONED_ISSUES_LINES in $(cat $MENTIONED_ISSUES); do
```
Iteration variable can not be plural. Please rename

romani replied:
MENTIONED_ISSUES should be rename to grep_lines

To mean that what content it has

Vyom-Yadav replied:
I feel `MENTIONED_ISSUES` is better than `grep_lines`, mentioned_issues describes the variables, grep is just a tool, changed it to `MENTIONED_ISSUES_LINE` i.e. singular form.

romani replied:
Please print content of MENTIONED_ISSUES_LINE here

romani replied:
@Vyom-Yadav, please do ∆∆

Vyom-Yadav replied:
Printed it.

romani replied:
From logs:
./config/import-control.xml:107:checkstyle/checkstyle/issues/3511

Vyom-Yadav replied:
Yes it is of this nature -
```
./pom.xml:709:checkstyle/checkstyle/issues/2160
./pom.xml:1505:checkstyle/checkstyle/issues/5770
./pom.xml:1567:checkstyle/checkstyle/issues/11446
./.semaphore/semaphore.yml:56:checkstyle/checkstyle/issues/9807
./.semaphore/semaphore.yml:74:checkstyle/checkstyle/issues/9248
```
file_name:line_number:issue_url_suffix

romani replied:
It is not mentioned issue, it is just output of grep tool.
So better to name variable to explain it's content.

Vyom-Yadav replied:
`grep_line` is vague, the line also passed through stream editor (sed), going by that logic we can even name it `sed_line`, if I looked it at in the future it would confuse me. The variable represents mentioned issue and where is it mentioned, if we want to include grep, we can maybe name it something like `MENTIONED_ISSUES_GREP_OUTPUT` or something of that form, what do you think?

romani replied:
MENTIONED_ISSUES_GREP_OUTPUT this works for me, I just want to put in name message that content is not simple, it is more like structure

romani replied:
MENTIONED_ISSUE_Structure
MENTIONing_package

Vyom-Yadav replied:
`MENTIONED_ISSUES_GREP_OUTPUT` sounds good, changed to that.

romani commented on .ci/no_old_refs.sh:
```diff
@@ -5,27 +5,67 @@ set -e
+grep -IPonr "[Tt]il[l]? #\d{1,5}" . \
+  | sed -e 's/:[a-zA-Z].*#/:checkstyle\/checkstyle\/issues\//' >> $MENTIONED_ISSUES
+
+# $LINKED_ISSUES need formatting before the are used
```
Please move this block above mentioned_issues

Vyom-Yadav replied:
Done.

romani commented on .ci/no_old_refs.sh:
```diff
@@ -5,27 +5,67 @@ set -e
+LINKED_ISSUES_MENTIONED=/tmp/linked_issues_mentioned
 API_GITHUB_PREFIX="https://api.github.com/repos"
 GITHUB_HOST="https://github.com"
+CHECKSTYLE_ISSUE_PREFIX="https:\/\/github.com\/checkstyle\/checkstyle\/issues\/"
```
CHECKSTYLE_ISSUES_URL

Vyom-Yadav replied:
Done.

romani commented on .ci/no_old_refs.sh:
```diff
@@ -5,27 +5,67 @@ set -e
-    echo "Following issues are mentioned in code to do something after they are closed:"
+    echo -ne "\nFollowing lines are referencing closed issues, such lines should be removed:\n\n"
     cat $CLOSED_ISSUES
+    if [ -f "$LINKED_ISSUES_MENTIONED" ]; then
```
This is independent validation why is it inner IF

Vyom-Yadav replied:
Yes it is, changed it.

romani commented on .ci/no_old_refs.sh:
```diff
@@ -5,27 +5,67 @@ set -e
   if [ "$STATE" = "closed" ]; then
-    echo "$GITHUB_HOST/$issue" >> $CLOSED_ISSUES
+    echo "$LINK" >> $CLOSED_ISSUES
+  elif [ ! -z "$LINKED_ISSUES" ]; then
```
It sounds like independent validation, they highly unlikely to overlap.
Should it be "if" instead of "elif"

Vyom-Yadav replied:
Was doing some minor branch optimization, but it wasn't right, closed issues can also be linked to the PR, changed it to be independent now.

romani reviewed (changes requested):
Items:

romani commented on .ci/no_old_refs.sh:
```diff
@@ -5,27 +5,76 @@ set -e
   fi
 done
 
+SHOULD_NORMALLY_EXIT=true
```
Exit_code = 0

exit Exist_code

romani reviewed (changes requested):
Item

romani reviewed (approved):
Ok to merge

strkkk commented on .ci/no_old_refs.sh:
```diff
@@ -3,29 +3,76 @@
+LINKED_ISSUES_FORMATTED=/tmp/linked_issues
+CHECKSTYLE_ISSUE_URL="https:\/\/github.com\/checkstyle\/checkstyle\/issues\/"
+
+# $LINKED_ISSUES need formatting before the are used
```
`before theY are used` ?

Vyom-Yadav replied:
I added a comment explaining formatting.

strkkk commented on .ci/no_old_refs.sh:
```diff
@@ -3,29 +3,76 @@
+
 if [ -f "$CLOSED_ISSUES" ]; then
-    echo "Following issues are mentioned in code to do something after they are closed:"
+    echo -ne "\nFollowing lines are referencing closed issues, such lines should be removed:\n\n"
```
to be honest, this looks ugly. It is not so easy to understand flags, and `\n`s in string look weird.
why dont make it like
```
echo
echo "message"
echo
```
?

same for if block below.

romani replied:
+1

Vyom-Yadav replied:
I removed `\n`, what is difficult to understand about-
```bash
if [ -f "$CLOSED_ISSUES" ]; 
```
`EXIT_STATUS` is also self-explanatory, I am fairly new to shell scripting, please tell me if there is a better way to do it.

strkkk replied:
@Vyom-Yadav I wrote comment about using `echo` and its flags. If condition is ok, but for echo it is difficult. Please get rid of echo flags. If you need to print empty line, use `echo` without arguments to make it clearer

Vyom-Yadav replied:
Got it, Done.

strkkk commented on .ci/no_old_refs.sh:
```diff
@@ -3,29 +3,76 @@
+
 if [ -f "$CLOSED_ISSUES" ]; then
-    echo "Following issues are mentioned in code to do something after they are closed:"
+    echo -ne "\nFollowing lines are referencing closed issues, such lines should be removed:\n\n"
```
`such lines should be removed` that is misleading, may be issue should be reopened or new one has to be created.

Vyom-Yadav replied:
I modified the message, please take a look.

strkkk replied:
wording lgtm now, but please split it to 2 sentences and 2 echos. This will also fix my comment https://github.com/checkstyle/checkstyle/pull/11443#discussion_r839979595

Vyom-Yadav replied:
Done

strkkk commented on .ci/no_old_refs.sh:
```diff
@@ -3,29 +3,76 @@
+CHECKSTYLE_ISSUE_URL="https:\/\/github.com\/checkstyle\/checkstyle\/issues\/"
+
+# $LINKED_ISSUES need formatting before the are used
+if [ ! -z "$LINKED_ISSUES" ]; then
```
I tihnk it is better to have this formatting in pipeline step where we get these links.
Also, it will be good to show input/output example or description what "formatting" means

Vyom-Yadav replied:
I added this as a step in the workflow, also added an explanatory comment.

strkkk reviewed (changes requested):
Please also address https://github.com/checkstyle/checkstyle/pull/11443#issuecomment-1073519369

Vyom-Yadav commented:
> Please also address https://github.com/checkstyle/checkstyle/pull/11443#issuecomment-1073519369

I had already added an example for that, please see the minor commit.

> Same for several occurrences of same issue in one file

All the lines will be highlighted, I removed the unique flag, so every issue (even if same) will be highlighted.

strkkk commented on .github/workflows/no_old_refs.yml:
```diff
@@ -3,28 +3,52 @@
+      #######################################################################
+      # Linked issues, if present, are received in the following format-
+      # 37,38
+      # We convert it into-
```
hyphen can be replaced by colon or removed. Same for first line of description

Vyom-Yadav replied:
Done

strkkk commented:
lgtm, please do not forget to remove test commit

Vyom-Yadav commented:
> lgtm, please do not forget to remove test commit

Will remove it after all reviewers are satisfied.

rnveach commented on .github/workflows/no_old_refs.yml:
```diff
@@ -3,28 +3,52 @@
 #
 # Workflow starts when:
 # 1) push on master branch
+# 2) pull request is opened or synchronized or reopened
```
When is a PR synchronized?

romani replied:
https://github.community/t/what-is-a-pull-request-synchronize-event/14784

>that happens when the source branch is updated.

So simple example is force push to GitHub of sources

romani commented:
@Vyom-Yadav , please remove test commit to let me merge

Vyom-Yadav commented on .ci/jsoref-spellchecker/whitelist.words:
```diff
@@ -830,6 +830,7 @@ modifierorder
 Mohnen
 mojohaus
 moks
+mondeja
```
Name of action we use to get the linked issues.

Vyom-Yadav commented:
> @Vyom-Yadav , please remove test commit to let me merge

@romani Done.
